cmd/compose/up.go

```diff
@@ -164,15 +163,29 @@ func runUp(ctx context.Context, streams api.Streams, backend api.Service, create
 		consumer = formatter.NewLogConsumer(ctx, streams.Out(), streams.Err(), !upOptions.noColor, !upOptions.noPrefix, upOptions.timestamp)
 	}
 
-	attachTo := services
+	var attachTo []string
 	if len(upOptions.attach) > 0 {
 		attachTo = upOptions.attach
 	}
 	if upOptions.attachDependencies {
-		attachTo = project.ServiceNames()
+		if err := project.WithServices(attachTo, func(s types.ServiceConfig) error {
+			if s.Attach == nil || *s.Attach {
+				attachTo = append(attachTo, s.Name)
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
 	}
 	if len(attachTo) == 0 {
-		attachTo = project.ServiceNames()
+		if err := project.WithServices(services, func(s types.ServiceConfig) error {
```

Comment:
I think this should now be an `else if` and pass the `IgnoreDependencies` policy for the traversal

i.e.

* If you're attaching dependencies, iterate w/ dependencies
  * if `attachTo` is empty, `WithServices` will iterate all services
  * Otherwise, it'll just do the targeted services & their deps
* Otherwise, only iterate the `services` being up'd (w/o deps)

Reply:
indeed. Updated to use a `set`, so we don't have to get into the possible execution paths to guess unicity is preserved :)